repair/repair.cc

```diff
@@ -944,9 +944,22 @@ static future<> repair_cf_range(repair_info& ri,
 // Comparable to RepairSession in Origin
 static future<> repair_range(repair_info& ri, const dht::token_range& range) {
     auto id = utils::UUID_gen::get_time_UUID();
-    return do_with(get_neighbors(ri.db.local(), ri.keyspace, range, ri.data_centers, ri.hosts), [&ri, range, id] (std::vector<gms::inet_address>& neighbors) {
+    auto neighbors = ri.neighbors.empty() ? get_neighbors(ri.db.local(), ri.keyspace, range, ri.data_centers, ri.hosts) : ri.neighbors[range];
+    return do_with(std::move(neighbors.all), std::move(neighbors.mandatory), [&ri, range, id] (auto& neighbors, auto& mandatory_neighbors) {
       auto live_neighbors = boost::copy_range<std::vector<gms::inet_address>>(neighbors |
                     boost::adaptors::filtered([] (const gms::inet_address& node) { return gms::get_local_gossiper().is_alive(node); }));
+      for (auto& node : mandatory_neighbors) {
+           auto it = std::find(live_neighbors.begin(), live_neighbors.end(), node);
+           if (it == live_neighbors.end()) {
```

Comment:
Assume there are n1,n2,n3.  n3 is the mandatory neighbor.
With you proposal:
When we calculate the live_neighbors,  n3 is down, so live_neighbors = n1, n2.
When we call  `if (std::any_of(mandatory_neighbors.begin(), mandatory_neighbors.end(), [] (const gms::inet_address& node) { return !gms::get_local_gossiper().is_alive(node); })
`
n3 is up. 

So the check will pass.

This is wrong, because we will repair without the  mandatory node n3.

Reply:
I don't see how n3 may switch from DOWN to UP from the local shard point of view after the `live_neghbors` calculation and the code in question - there isn't any preemption point there. 

What do I miss?

Reply:
It does not have now but we can have it later. In order to prevent stalls, we have code like below
```
std::copy_if(x.begin(), x.end(), std::inserter(set_diff, set_diff.end()),
        [&y] (auto& item) { thread::maybe_yield(); return y.find(item) == y.end(); });
```
to prevent a long operation stalls the reactor.

We might apply similar trick to live_neighbors calculation as well and forget the code depends on it does not yield to work correctly.

Reply:
Sounds like a very long shot but ok...